synthetics-stats/src/entities/userBalance.ts

```diff
@@ -0,0 +1,127 @@
+import { BigInt, log } from "@graphprotocol/graph-ts";
+import {
+  CollectedMarketFeesInfo,
+  LatestUserGmTokensBalanceChangeRef,
+  Transaction,
+  UserGmTokensBalanceChange
+} from "../../generated/schema";
+import { getOrCreateCollectedMarketFees } from "./fees";
+
+let ZERO = BigInt.fromI32(0);
+
+export function saveUserGmTokensBalanceChange(
+  account: string,
+  marketAddress: string,
+  value: BigInt,
+  transaction: Transaction,
+  transactionLogIndex: BigInt,
+  postfix: string
+): void {
+  let prevEntity = getLatestUserGmTokensBalanceChange(account, marketAddress);
+  let entity = _createUserGmTokensBalanceChange(account, marketAddress, transaction, transactionLogIndex, postfix);
+  let totalFees = CollectedMarketFeesInfo.load(marketAddress + ":total");
+  let prevBalance = prevEntity ? prevEntity.tokensBalance : ZERO;
+  let prevCumulativeIncome = prevEntity ? prevEntity.cumulativeIncome : ZERO;
+  let income = prevEntity ? calcIncomeForEntity(prevEntity) : ZERO;
+  let isDeposit = value.gt(ZERO);
+
+  entity.tokensBalance = prevBalance.plus(value);
+  entity.cumulativeIncome = prevCumulativeIncome.plus(income);
+
+  if (totalFees) {
+    entity.cumulativeFeeUsdPerGmToken = isDeposit
+      ? // These transfers happen during DepositExecuted and WithdrawalCreated transactions (different stages)
+        // That's why GM transfer goes After fee is received.
+        // And withdrawal GM balance changes Before fee is received.
+        // so for deposits we are storing extra (prev) fee, and act normally for withdrawals
+        totalFees.prevCumulativeFeeUsdPerGmToken
+      : totalFees.cumulativeFeeUsdPerGmToken;
+
+    // we need this on UI side to properly restore income per balance change
+    // (regardless of comment above, just always actual fee by that moment)
+    entity.actualCumulativeFeeUsdPerGmToken = totalFees.cumulativeFeeUsdPerGmToken;
+  }
+  entity.index = getBalanceChangeNextIndex(account, marketAddress);
+
+  entity.save();
+
+  saveLatestUserGmTokensBalanceChange(entity);
+}
+
+function getBalanceChangeNextIndex(account: string, marketAddress: string): BigInt {
+  let id = account + ":" + marketAddress;
+  let latestRef = LatestUserGmTokensBalanceChangeRef.load(id);
+
+  return latestRef ? latestRef.nextIndex : BigInt.fromI32(0);
+}
+
+function getLatestUserGmTokensBalanceChange(account: string, marketAddress: string): UserGmTokensBalanceChange | null {
+  let id = account + ":" + marketAddress;
+  let latestRef = LatestUserGmTokensBalanceChangeRef.load(id);
+
+  if (!latestRef) return null;
+
+  let latestId = latestRef.latestUserGmTokensBalanceChange;
+
+  if (latestId) return UserGmTokensBalanceChange.load(latestId);
```

Comment:
can `latestId` be null here? at first glance if it exists than it always has `latestUserGmTokensBalanceChange`

Reply:
if it's a sanity check it probably makes sense to throw if it's null?